### PR TITLE
recipe for Shotcut video editor

### PR DESCRIPTION
**Pull request by TURX**
This is for #4028.

### CODE REVIEW

**Comment by TURX** [comment]
I have got an error in building libmlt as described at https://paste.ubuntu.com/p/nXp6rGqckP/. How could I figure it out?

**Comment by korli** [comment]
The linker by default builds an executable, which needs a main function. Did you try to patch https://github.com/mltframework/mlt/blob/master/configure#L97 ?

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 1]
```diff
@@ -0,0 +1,75 @@
+SUMMARY="A free, open source, cross-platform video editor"
+DESCRIPTION="Major features include support for a wide range of formats; no import required meaning native timeline editing; Blackmagic Design support for input and preview monitoring; and resolution support to 4k."
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]: please wrap to 80 chars.

**Comment by threedeyes** [comment]
It works?
I built it last year, but Shotcut required OpenGL support in Qt for rendering. At the moment, I have not implemented this functionality in QPA (platform plugin). The same situation with Kdenlive.

**Comment by TURX** [comment]
> The linker by default builds an executable, which needs a main function. Did you try to patch https://github.com/mltframework/mlt/blob/master/configure#L97 ?

After patching this, the build still does not work with a similar error: https://paste.ubuntu.com/p/pJkrZSwtQM/.

**Comment by Begasus** [comment]
Could you create a separate PR for MLT? It's confusing to use 2 recipes in 1 PR

File: media-libs/mlt/mlt-6.18.0.recipe [hunk 2]
```diff
@@ -0,0 +1,63 @@
+SUMMARY="An open source multimedia framework"
+DESCRIPTION="It provides a toolkit for video editors and many more types of applications."
```
Comment (Begasus) [inline media-libs/mlt/mlt-6.18.0.recipe]: Line is still over 80 chars, check the guidelines: https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines

**Comment by Begasus** [comment]
> > The linker by default builds an executable, which needs a main function. Did you try to patch https://github.com/mltframework/mlt/blob/master/configure#L97 ?
> 
> After patching this, the mlt build still does not work with a similar error: https://paste.ubuntu.com/p/pJkrZSwtQM/.

For this you need to patch mlt++/configure

**Comment by TURX** [comment]
The modification of MLT is separated in #4608.

**Comment by Begasus** [comment]
> The modification of MLT is separated in #4608.

Thanks, could you remove the mlt files from this PR?

**Comment by Begasus** [comment]
This is a wip, deppendency https://github.com/haikuports/haikuports/pull/4608

**Comment by TURX** [comment]
There is no devel:libqt5x11extras found in Haiku. Should I port this?

**Comment by Begasus** [comment]
Try to stay away from X11 if you can

**Comment by TURX** [comment]
> Try to stay away from X11 if you can

It is defined at line 368 of /src/src.pro. How could I use two exceptions (mac and haiku) here?

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	cmd:lrelease$secondaryArchSuffix >= 5
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:qmake
+	"
+
+BUILD()
+{
+	qmake PREFIX=$prefix
+	make -j$jobs
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	make $jobArgs
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	frei0r${secondaryArchSuffix}_devel
+	ladspa_sdk${secondaryArchSuffix}_devel
+	"
+BUILD_PREREQUIRES="
+	cmd:gcc$secondaryArchSuffix
+	cmd:getconf
+	cmd:lrelease$secondaryArchSuffix >= 5
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:qmake
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	cmd:qmake >= 5
```

**Comment by TURX** [comment]
Although I have modified the patchset and the recipe, the build still fails: https://paste.ubuntu.com/p/7MvvkC5Mj4/.
I have imported mlt-related things, but there is no include-path for mlt implied from the log.

File: media-video/shotcut/patches/shotcut-19.12.31.patchset [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+     build_pass:CONFIG(debug, debug|release) {
+@@ -365,7 +365,7 @@ win32 {
+     }
+     RC_FILE = ../packaging/windows/shotcut.rc
+ }
+-unix:!mac {
++unix:!mac:!haiku {
+     QT += x11extras
+     CONFIG += link_pkgconfig
+     PKGCONFIG += mlt++
```
Comment (korli) [inline media-video/shotcut/patches/shotcut-19.12.31.patchset]: CONFIG and PKGCONFIG should be kept for unix:!mac

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	shotcut$secondaryArchSuffix = $portVersion
+ 	app:shotcut
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	cmd:lame
+	lib:frei0r$secondaryArchSuffix
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	frei0r$secondaryArchSuffix
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	lib:ladspa_sdk$secondaryArchSuffix
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	ladspa_sdk$secondaryArchSuffix
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+BUILD()
+{
+	qmake PREFIX=$prefix INCLUDEPATH=$includeDir
+	make $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir/Shotcut
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	mkdir -p $appsDir
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	cmd:getconf
+	cmd:lrelease$secondaryArchSuffix >= 5
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:qmake >= 5
+	"
+
+BUILD()
+{
+	qmake PREFIX=$prefix INCLUDEPATH=$includeDir
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	qmake PREFIX=$prefix
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	make $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir/Shotcut
+	make -j$jobs install
+	cp -rf ../bin/app/* $appsDir/Shotcut
+	cp app/Shotcut $appsDir/Shotcut
+	addAppDeskbarSymlink $appsDir/Shotcut/Shotcut
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	addAppDeskbarSymlink $appsDir/Shotcut
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+{
+	qmake PREFIX=$prefix INCLUDEPATH=$includeDir
+	make $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir/Shotcut
+	make -j$jobs install
+	cp -rf ../bin/app/* $appsDir/Shotcut
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	mv $prefix/bin/shotcut $appsDir
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+BUILD()
+{
+	qmake PREFIX=$prefix
+	make VERBOSE=1 $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir
+	make VERBOSE=1 -j$jobs install
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	make install
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+	cp -rf ../bin/app/* $appsDir
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
```

File: media-video/shotcut/patches/shotcut-19.12.31.patchset (hunk 5 above)
Comment (korli) [reply media-video/shotcut/patches/shotcut-19.12.31.patchset]: This is the reason why mlt header path is not picked up.

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	devel:libQt5Network$secondaryArchSuffix
+	devel:libQt5Quick$secondaryArchSuffix
+	devel:libQt5Test$secondaryArchSuffix
+	devel:libQt5WebKitWidgets$secondaryArchSuffix
+	devel:libQt5Xml$secondaryArchSuffix
+	devel:libwebm$secondaryArchSuffix
+	devel:libx264$secondaryArchSuffix
+	frei0r${secondaryArchSuffix}_devel
+	ladspa_sdk${secondaryArchSuffix}_devel
+	mlt${secondaryArchSuffix}_devel
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+BUILD()
+{
+	qmake PREFIX=$prefix
+	make $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir
+	make -j$jobs install
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	make install
```

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+	lib:libQt5WebKitWidgets$secondaryArchSuffix
+	lib:libQt5Xml$secondaryArchSuffix
+	lib:libwebm$secondaryArchSuffix
+	lib:libx264$secondaryArchSuffix
+	mlt$secondaryArchSuffix
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	cmd:ffmpeg
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
```
cmd:ffmpeg should be a runtime dependency

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	lib:libQt5Designer$secondaryArchSuffix
+	lib:libQt5Gui$secondaryArchSuffix
+	lib:libQt5Network$secondaryArchSuffix
+	lib:libQt5Quick$secondaryArchSuffix
+	lib:libQt5Test$secondaryArchSuffix
+	lib:libQt5WebKitWidgets$secondaryArchSuffix
+	lib:libQt5Xml$secondaryArchSuffix
+	lib:libwebm$secondaryArchSuffix
+	lib:libx264$secondaryArchSuffix
+	mlt$secondaryArchSuffix
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
```

**Comment by korli** [comment]
Please also adjust the melt binary name in src/jobs/meltjob.cpp line 91 and add a runtime dependency for cmd:melt in the recipe.

File: media-video/shotcut/patches/shotcut-19.12.31.patchset [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+     LIBS += -lX11
+ }
++haiku {
```
Comment (korli) [inline media-video/shotcut/patches/shotcut-19.12.31.patchset]:
```suggestion
+unix:!mac {
```
then remove the duplicate lines in the block unix:!mac:!haiku

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+	qmake PREFIX=$prefix
+	make $jobArgs
+}
+
+INSTALL()
+{
+	mkdir -p $appsDir
+	make install
+	mv $prefix/bin/shotcut $appsDir
+	addAppDeskbarSymlink $appsDir/shotcut
```
Comment (korli) [inline media-video/shotcut/shotcut-19.12.31.recipe]:
```suggestion
	addAppDeskbarSymlink $appsDir/shotcut Shotcut
```

the symlink could so be capitalized.

**Comment by TURX** [comment]
I have another issue:
```
Warning: POLICY ERROR: Invalid top-level package entry "share"
```

**Comment by korli** [comment]
> What is the alternative of $$PREFIX/share in src.pro?

you can move the content of share in $dataDir after "make install", and adjust the directory in qmlutilities.cpp

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+	frei0r${secondaryArchSuffix}_devel
+	ladspa_sdk${secondaryArchSuffix}_devel
+	"
+BUILD_PREREQUIRES="
+	cmd:gcc$secondaryArchSuffix
+	cmd:getconf
+	cmd:lrelease$secondaryArchSuffix >= 5
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:qmake >= 5
```
Comment (Begasus) [inline media-video/shotcut/shotcut-19.12.31.recipe]: Needs secondaryArchSuffix
```diff
- 	cmd:qmake >= 5
+ 	cmd:qmake$secondaryArchSuffix >= 5
```

**Comment by TURX** [comment]
> Please also adjust the melt binary name in src/jobs/meltjob.cpp line 91 and add a runtime dependency for cmd:melt in the recipe.

What is the meltPath for Haiku?

**Comment by korli** [comment]
> > Please also adjust the melt binary name in src/jobs/meltjob.cpp line 91 and add a runtime dependency for cmd:melt in the recipe.
> 
> What is the meltPath for Haiku?

I would just replace "qmelt" with "melt". Maybe the shotcut binary should stay in /bin for this reason.

**Comment by TURX** [comment]
> > What is the alternative of $$PREFIX/share in src.pro?
> 
> you can move the content of share in $dataDir after "make install", and adjust the directory in qmlutilities.cpp

Do you mean that I should adjust the directory to $dataDir?

**Comment by TURX** [comment]
It builds successfully to a hpkg package: https://paste.ubuntu.com/p/rTGxgBTM8w/.

**Comment by TURX** [comment]
The desktop link is broken, and I would fix this issue in rev 12.

**Comment by TURX** [comment]
Also, JACK is required:



Should I port it?

**Comment by TURX** [comment]
Here is the output of the latest version:

```
~/haikuports> /boot/system/apps/shotcut 
[Info   ] <Application::Application> Starting Shotcut version 20.01.20 
[Info   ] <Application::Application> Linux version 
[Info   ] <Application::Application> number of logical cores = 4 
[Info   ] <Application::Application> locale = QLocale(Japanese, Japanese, Japan) 
[Info   ] <Application::Application> install dir = "/boot/system/apps" 
[Info   ] <Application::Application> device pixel ratio = 1 
[Debug  ] <ShotcutSettings::log> language "ja_JP" 
[Debug  ] <ShotcutSettings::log> deinterlacer "onefield" 
[Debug  ] <ShotcutSettings::log> external monitor "" 
[Debug  ] <ShotcutSettings::log> GPU processing false 
[Debug  ] <ShotcutSettings::log> interpolation "bilinear" 
[Debug  ] <ShotcutSettings::log> video mode "" 
[Debug  ] <ShotcutSettings::log> realtime true 
[Debug  ] <ShotcutSettings::log> audio channels 2 
[Debug  ] <MainWindow::changeTheme> begin 
[Debug  ] <MainWindow::changeTheme> end 
[Warning] <> QObject::~QObject: Timers cannot be stopped from another thread
```

**Comment by TURX** [comment]
> Why are libQt5Sql and sqlite3 needed?

https://paste.ubuntu.com/p/ZQMN2nrQs6/ provided by Begasus at 00:50 China Standard Time in #haiku IRC.

**Comment by korli** [comment]
> > Why are libQt5Sql and sqlite3 needed?
> 
> https://paste.ubuntu.com/p/ZQMN2nrQs6/ provided by Begasus at 00:50 China Standard Time in #haiku IRC.

At least drop the dependency on devel:libsqlite3.

**Comment by korli** [comment]
also please restore cmd:melt  cmd:ffmpeg: shotcut runs these tools directly.
see https://github.com/mltframework/shotcut/blob/master/src/jobs/ffmpegjob.cpp
https://github.com/mltframework/shotcut/blob/master/src/jobs/meltjob.cpp

**Comment by TURX** [comment]
After building and installing jack2, Shotcut seems to be loaded JACK and then crashed. No SQL-related error happened.

Terminal log: https://paste.ubuntu.com/p/Srqr26mztt/
Error report: https://paste.ubuntu.com/p/NYcR3dFYKf/

Should I disable the pthread?

**Comment by korli** [comment]
mlt should depend on sdl or sdl2, so that a consumer is available.

**Comment by TURX** [comment]
> mlt should depend on sdl or sdl2, so that a consumer is available.

Should I also port sdl or sdl2 or you could just rest this PR or enable me to pick up another task and I work on this PR in the meantime?

**Comment by korli** [comment]
sdl or sdl2 are already ported. As @Begasus and @threedeyes  said, Shotcut probably should stay disabled until OpenGL in Qt is fixed. you can disable architectures with "?", and the task can be approved.

**Comment by TURX** [comment]
The same error happens after importing libsdl2_2.0. The error report does not change when the lib is libsdl.

Error report: https://paste.ubuntu.com/p/vQSNF9qFbq/

**Comment by TURX** [comment]
Now the recipe is disabled.

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 21]
```diff
@@ -9,8 +9,8 @@ CHECKSUM_SHA256="d7017d605edb581cb47de4844827bd627c159c810614381f4586ef35e8c3d97
 SOURCE_DIR="shotcut-$portVersion"
 PATCHES="shotcut-$portVersion.patchset"
 
-ARCHITECTURES="x86_64"
-SECONDARY_ARCHITECTURES="x86"
+ARCHITECTURES="!x86_64"
+SECONDARY_ARCHITECTURES="!x86"
```
Comment (Begasus) [inline media-video/shotcut/shotcut-19.12.31.recipe]: Mark disabled not as broken as suggested by @korli 
```diff
-ARCHITECTURES="!x86_64"
-SECONDARY_ARCHITECTURES="!x86"
+ARCHITECTURES="?x86_64"
+SECONDARY_ARCHITECTURES="?x86"
```

**Comment by Begasus** [comment]
> sdl or sdl2 are already ported. As @Begasus and @threedeyes said, Shotcut probably should stay disabled until OpenGL in Qt is fixed. you can disable architectures with "?", and the task can be approved.

You can build/add a module for SDL and SDL2 for mlt, shouldn't be needed here

**Comment by Begasus** [comment]
Current log when trying to launch Shotcut from Terminal (with mlt builtin modules for SDL/SDL2 ...)
[shotcut_log.txt](https://github.com/haikuports/haikuports/files/4089446/shotcut_log.txt)

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+	devel:libQt5Network$secondaryArchSuffix
+	devel:libQt5Quick$secondaryArchSuffix
+	devel:libQt5Sql$secondaryArchSuffix
+	devel:libQt5Test$secondaryArchSuffix
+	devel:libQt5WebKitWidgets$secondaryArchSuffix
+	devel:libQt5Xml$secondaryArchSuffix
+	devel:libwebm$secondaryArchSuffix
+	devel:libx264$secondaryArchSuffix
+	frei0r${secondaryArchSuffix}_devel
+	ladspa_sdk${secondaryArchSuffix}_devel
```
Comment (Begasus) [inline media-video/shotcut/shotcut-19.12.31.recipe]: Move these 2 lines below haiku${secondaryArchSuffix}_devel

File: media-video/shotcut/shotcut-19.12.31.recipe [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+	make install
+
+	mkdir -p $appsDir $dataDir
+	mv $prefix/share/shotcut $dataDir
+	mv $prefix/share/metainfo $dataDir
+	mv $prefix/share/applications $dataDir
+	mv $prefix/share/mime/packages $dataDir
+	mv $prefix/share/icons/hicolor/64x64/apps $dataDir
+	mv $prefix/share/man/man1 $dataDir
+	rm -rf $prefix/share
```
Comment (Begasus) [inline media-video/shotcut/shotcut-19.12.31.recipe]: applications and icons shouldn't be needed and can be removed
man1 should go to $manDir

**Pull request merged**